### ISSUE TITLE
Where does portmaster place itself relative to distro firewalls (firewalld/ufw)?

### ISSUE THREAD

**Issue by Arcitec** [safing/portmaster#801]
Hello Safing team! ❤️ Thank you so much for creating Portmaster. I haven't been this excited about new software in years. It is fantastic.

I am curious how the portmaster rules execute in relation to the OS/distro firewall rules on Linux?

For example, this is how firewalld works on Fedora Workstation:

- Deny all incoming connections.
- Allow incoming on ports 1025-65535.
- Allow certain incoming services such as mdns and ssh.

What happens when Portmaster is installed on a system that uses `firewalld`?

My **guess (or at least hope)** is this:

- Portmaster places itself at the top as the highest priority rule.
- Unmarked connections are forwarded to the Portmaster user space for classification.
- All packets for that connection are then marked as either accept or drop, by the flag that Portmaster decided.
- Any packet that is marked as accept then goes to the remaining rules which `firewalld` created, which block all incoming connections but then allow specific service ports (ssh, 1025-65535, etc).

If this is what happens, then I would be able to set Portmaster as "allow all incoming" to basically just forward everything to the rules that are already set up by Fedora in their `firewalld`. That would let me use Portmaster for my intended purpose: Per app blocking of outgoing connections. I would also be able to block specific app's incoming connections to get extra granularity.

In short if it works as I hope it does (portmaster processing its rules first, and marking the connections, and then handing off to the other ufw/firewalld rules in iptables), then it would be the best of all worlds.

**Comment by Bananaman** [comment]
@p2anav Hi! Thank you for answering. However, this was a Linux question. There is no Portmaster kernel extension on Linux, since I read the docs already where you mention using the built-in Linux nftables:

https://docs.safing.io/portmaster/architecture/os-integration#linux

My question is about how the Portmaster rules interact with other firewalls, specifically what happens to my existing rules in the Linux kernel. The original question mentions an example.

If you use Linux, perhaps you know how that works? Or perhaps @dhaavi knows since I noticed he often replies to Linux related tickets? 😊

**Comment by Bananaman** [comment]
I tried to investigate this.

I noticed that ever since installing portmaster,  the "firewalld" service status is "inactive (dead)" and none of its rules are in iptables.

So the answer seems to be that it totally replaces firewalld.

In fact I totally discovered why.

In "portmaster.service" it is marked as "Conflicts=firewalld.service".

So as long as portmaster is enabled, then system auto-disables firewalld.

# That is totally fine behavior.  I just needed to know it!

So you can update the "Compatibility" site to say that Firewalld and all of its rules are disabled while the portmaster service is enabled.

And that if a person wants to allow incoming connections to various services that were set up in firewalld, then I recommend that they install "firewall-config" (the official GUI) and make notes about what services are enabled and what ports and networks they use, and then manually write equivalent rules in Portmaster.

I will be migrating Fedora Workstation's default firewalld rules into Portmaster.

**Comment by itsnotsaved** [comment]
>I will be migrating Fedora Workstation's default firewalld rules into Portmaster.

@Bananaman When portmaster running, I don't have a problem with https://www.configserver.com/cp/csf.html Maybe, you can use it?

@portmaster-team, don't make this tool incompatible with csf too 🤣

**Comment by CodingMadness** [comment]
> > I will be migrating Fedora Workstation's default firewalld rules into Portmaster.
> 
> [@Bananaman](https://github.com/Bananaman) When portmaster running, I don't have a problem with [configserver.com/cp/csf.html](https://www.configserver.com/cp/csf.html) Maybe, you can use it?
> 
> [@safing](https://github.com/safing) team, don't make this tool incompatible with csf too 🤣

May I know, is this already implemented on the kernel-level or did it get abandoned?
